File: MATLAB/IM.m
function [C,c] = IM(par)
%IM utilises an interpolation-based mapping to solve the 1D ADR equation.
%
% IM(par) assumes par is a struct array containing information about the 
% transport problem. 

% Transport coefficients and problem parameters
N = par.N;

M = par.M;

xF = par.xF(:);
xC = par.xC(:);

t0 = par.t0;
tK = par.tK;
K = par.K;
tau = (tK - t0) / K;
t = t0:tau:tK;

R = par.R;
D = par.D;
v = par.v;
mu = par.mu;
Gamma = par.Gamma;

omega = par.omega;

alpha0 = par.alpha0;
beta0 = par.beta0;
sigma0 = par.sigma0;

alphaL = par.alphaL;
betaL = par.betaL;
sigmaL = par.sigmaL;

g0 = par.g0(t);
gL = par.gL(t);

C = zeros(M,K+1);
C(:,1) = par.c0(xC);

c = zeros(N,K+1);
c(:,1) = par.c0(xF);

I = eye(N,N);

% Construct the mapping matrix G [Equation 13]
G = interpG(xF,xC);

% Construct the iteration matrix A [Equation 7] and vector b [Equation 8]
[A,b] = fvmStd(R,D,v,mu,Gamma,omega,alpha0,beta0,sigma0,alphaL,betaL,sigmaL,xF);

b0 = b(1);
bL = b(N);

b(2:N-1) = tau*b(2:N-1);

val0 = Gamma(xF(1)) / R(xF(1));
valL = Gamma(xF(N)) / R(xF(N));

itMat = G' * (I - tau*A) * G;

% Time stepping
for kk = 1:K

    b(1) = tau * (b0*g0(kk+1) + val0);

    b(N) = tau * (bL*gL(kk+1) + valL);

    % Approximate solution on the coarse grid [Equation 15]
    C(:,kk+1) = itMat \ ( G' * ( c(:,kk) + b ) );

    % Reconstruct solution on the fine grid [Equation 14]
    c(:,kk+1) = G * C(:,kk+1);

end

end

File: Julia/IM.jl
using LinearAlgebra

include("fvmStd.jl")

# Construct mapping matrix by linear interpolation [Equation 13]
function interpG(N,xF,M,xC)

	G = zeros(N,M)
	P = div(N-1,M-1) + 1

	for m in 1:M-1

		idx = (m-1)*(P-1)+1:m*(P-1)+1

		G[idx,m] .= (xC[m+1] .- xF[idx]) ./ (xC[m+1] .- xC[m])
		G[idx,m+1] .= (xF[idx] .- xC[m]) ./ (xC[m+1] .- xC[m])

	end

	return G

end

function IM(par)

	# Transport coefficients and problem parameters
	x0 = par.x0
	xL = par.xL

	N = par.N
	M = par.M

	xF = par.xF
	xC = par.xC

	tau = par.tau
	t = par.t
	K = par.K

	R = par.R
	D = par.D
	v = par.v
	mu = par.mu
	Gamma = par.Gamma

	omega = par.omega

	alpha0 = par.alpha0
	beta0 = par.beta0
	sigma0 = par.sigma0
	g0 = par.g0(t)

	alphaL = par.alphaL
	betaL = par.betaL
	sigmaL = par.sigmaL
	gL = par.gL(t)

	# Construct the mapping matrix G [Equation 13]
	G = interpG(N,xF,M,xC)
	GT = transpose(G)

	# Construct the iteration matrix A [Equation 7] and vector b [Equation 8]
	A, b = fvmStd(xF,R,D,v,mu,Gamma,omega,alpha0,beta0,sigma0,alphaL,betaL,sigmaL)
	itMatFull = I - tau*A

	# Pre-allocate the b vectors for use in the time-stepping loop
	bval = zeros(N,K+1)
	bval[1,:] .= tau .* (b[1] .* g0 .+ Gamma(x0) / R(x0))
	bval[N,:] .= tau .* (b[N] .* gL .+ Gamma(xL) / R(xL))
	@views @inbounds for k in 1:K
		bval[2:N-1,k] .= tau .* b[2:N-1]
	end

	# Construct the coarse-grid iteration matrix [Equation 15]
	itMatCoarse = GT * itMatFull * G
	itMatCoarse = lu(itMatCoarse)

	# Pre-allocate full Nx1 solutions
	c = zeros(N,K+1)
	c[:,1] = par.c0(xF)

	# Pre-allocate coarse Mx1 solutions
	C = zeros(M,K+1)
	C[:,1] = par.c0(xC)

	# Pre-allocate RHS vectors for the full (stepTmpbFull) NxN problem, and the coarse (stepTmpbCoarse) MxM problem
	stepTmpbFull = similar(xF)
	stepTmpbCoarse = similar(xC)

	# Time stepping
	for k in 1:K

		# Construct the full Nx1 RHS vector
		stepTmpbFull .= c[:,k]
		stepTmpbFull .+= bval[:,k+1]

		# Construct the coarse Mx1 RHS vector
		stepTmpbCoarse .= GT * stepTmpbFull

		# Approximate solution on the coarse grid [Equation 15]
		ldiv!(itMatCoarse,stepTmpbCoarse)
		C[:,k+1] .= stepTmpbCoarse

		# Reconstruct solution on the fine grid [Equation 14]
		c[:,k+1] .= G * C[:,k+1]

	end

	return c, C

end